db/db_impl/db_impl_write.cc

```diff
@@ -1835,8 +1835,26 @@ Status DB::Put(const WriteOptions& opt, ColumnFamilyHandle* column_family,
 
 Status DB::Delete(const WriteOptions& opt, ColumnFamilyHandle* column_family,
                   const Slice& key) {
-  WriteBatch batch;
-  batch.Delete(column_family, key);
+  if (nullptr == opt.timestamp) {
+    WriteBatch batch;
+    Status s = batch.Delete(column_family, key);
+    if (!s.ok()) {
+      return s;
+    }
+    return Write(opt, &batch);
+  }
+  const Slice* ts = opt.timestamp;
+  assert(ts != nullptr);
+  const size_t ts_sz = ts->size();
+  WriteBatch batch(key.size() + ts_sz + 24, /*max_bytes=*/0, ts_sz);
```

Comment:
It can be found in https://github.com/facebook/rocksdb/blob/5.14.fb/db/db_impl_write.cc#L1461:L1463.
`WriteBatch` starts with a header `WriteBatchInternal::kHeader` which is 12 bytes (8-byte seq and 4-byte count). Then we have one byte tag for `Put` operation. We also have 12 bytes for key length + value length (together) Var-int encoded. The total is 8 + 4 + 1 + 11 = 24.

Reply:
I'm wondering if we could use (compile-time) constants that explain this, e.g. something like:

```
constexpr size_t write_batch_overhead = WriteBatchInternal::kHeader + sizeof(ValueType) + ...; // is there a pre-existing constant that can be put here?
```

or

```
constexpr size_t key_and_value_length_size = 11;
constexpr size_t write_batch_overhead = WriteBatchInternal::kHeader + sizeof(ValueType) + key_and_value_length_size;
```

Or if all else fails :)

```
constexpr size_t write_batch_overhead = 24;
```